src/main/scala/calc/Compiler.scala

```diff
@@ -74,8 +99,72 @@ object Compiler {
       case Literal(value) =>
         irt.DoubleLiteral(value)
 
-      case _ =>
-        throw new Exception(
+      case BinaryOp(op, lhs, rhs) => {
+        val jsop = op match {
+          case "+" => irt.BinaryOp.Double_+
+          case "-" => irt.BinaryOp.Double_-
+          case "*" => irt.BinaryOp.Double_*
+          case "/" => irt.BinaryOp.Double_/
+        }
+        irt.BinaryOp(jsop, irt.Unbox(compileExpr(lhs)(env, typeEnv),'D'), irt.Unbox(compileExpr(rhs)(env, typeEnv),'D'))
+      }
+
+      case Let(ident, value, body) => {
+        /*var jsV:irt.Tree = if (value.isInstanceOf[Closure]){
+          compileClosure(value.asInstanceOf[Closure], ident.name)(env, typeEnv)
+        } else{
+          compileExpr(value)(env, typeEnv)
+        }*/
+        val jsV = compileExpr(value)(env, typeEnv)
+        val jsIdent = irt.Ident(ident.name)
+        val jsId = irt.VarDef(jsIdent, jsV.tpe, false, jsV)
+        irt.Block(List(jsId, compileExpr(body)(env+(ident.name -> jsV), typeEnv+(ident.name -> jsV.tpe))))
+      }
+
+      case Ident(n) => {
+        try {
+          irt.VarRef(irt.Ident(n))(typeEnv(n))
+        }catch{
+          case e:Exception => irt.VarRef(irt.Ident(n))(irtpe.AnyType)
+          //throw new Exception(s"The variable ${n} is not in the scope")
+        }
+      }
+
+      case If(cond, thenp, elsep) => {
+        val jsThenp = compileExpr(thenp)(env, typeEnv)
+        val jsElsep = compileExpr(elsep)(env, typeEnv)
+        val jsCond = compileExpr(cond)(env, typeEnv)
+        val finalCond = irt.BinaryOp(irt.BinaryOp.!==, irt.Unbox(jsCond,'D'), irt.DoubleLiteral(0.0))
+        if (jsElsep.tpe != jsElsep.tpe){
+          throw new Exception(s"The then and also phase in the conditional statement must bse same type")
+        }
+        irt.If(finalCond, jsThenp, jsElsep)(jsThenp.tpe)
+      }
+
+      case Closure(idList, body) => {
+        /*val capturePara = idList.map(s => s match{
+          case Ident(n) => irt.ParamDef(irt.Ident(n), irtpe.AnyType, false, false)
+          case _ => throw new Exception("Function def error!")
+        })*/
+        val captureSet = findCapture(tree)
+        val captureParam = captureSet.map(s => irt.ParamDef(irt.Ident(s), typeEnv(s), false, false)).toList
+        val params = idList.map(s => irt.ParamDef(irt.Ident(s.name), irtpe.AnyType, false, false))
+        irt.Closure(captureParam, params, compileExpr(body)(env, typeEnv), captureSet.map(s => env(s)).toList)
```

Comment:
`s => env(s)` is duplicating the rhs (right-hand-side) of `let` expressions if they are captured. You should simply use a `irt.VarRef` to the variable declared by the `let` expression.